Geometry/FindIntersections.js

```diff
@@ -0,0 +1,199 @@
+ * @param {Segment} seg1 - First segment.
+ * @param {Segment} seg2 - Second segment.
+ * @param {Array<Event>} events - Array of events to update.
+ * @param {Array<Intersection>} intersections - Array of intersections to update.
+ */
+export const handlePotentialIntersection = (
+  seg1,
+  seg2,
+  events,
+  intersections
+) => {
+  const intersectionPoint = calculateIntersectionPoint(seg1, seg2)
+  if (intersectionPoint) {
+    events.push(createEvent(intersectionPoint, ['intersection'], [seg1, seg2]))
+    intersections.push({
+      segment1: seg1,
+      segment2: seg2,
+      point: intersectionPoint
+    })
+  }
+}
+
+/**
+ * Finds all intersections among a set of line segments using the Bentley-Ottmann algorithm.
+ * @param {Array<Segment>} segments - Array of line segments to check for intersections.
+ * @returns {Array<Intersection>} Array of all intersections found.
+ */
+const findIntersections = (segments) => {
+  const events = []
+  const intersections = []
+  let sweepLineStatus = []
+
+  // Create initial events
+  segments.forEach((segment) => {
+    events.push(createEvent(segment.start, ['left'], [segment]))
+    events.push(createEvent(segment.end, ['right'], [segment]))
+  })
+
+  // Sort events
+  events.sort((a, b) => {
+    const pointCompare = comparePoints(a.point, b.point)
+    if (pointCompare !== 0) return pointCompare
+
+    // If points are the same, prioritize: intersection > left > right
+    const typePriority = { intersection: 0, left: 1, right: 2 }
+    return (
+      Math.min(...a.types.map((t) => typePriority[t])) -
+      Math.min(...b.types.map((t) => typePriority[t]))
+    )
+  })
+
+  // Process events
+  events.forEach((event) => {
+    if (event.types.includes('left')) {
+      event.segments.forEach((segment) => {
+        sweepLineStatus.push(segment)
+        sweepLineStatus.sort(compareSegmentsY)
+        const index = sweepLineStatus.indexOf(segment)
+        const lower = index > 0 ? sweepLineStatus[index - 1] : null
+        const upper =
+          index < sweepLineStatus.length - 1 ? sweepLineStatus[index + 1] : null
+        if (lower)
+          handlePotentialIntersection(segment, lower, events, intersections)
+        if (upper)
+          handlePotentialIntersection(segment, upper, events, intersections)
+      })
+    }
+
+    if (event.types.includes('right')) {
+      event.segments.forEach((segment) => {
+        const index = sweepLineStatus.indexOf(segment)
+        const lower = index > 0 ? sweepLineStatus[index - 1] : null
+        const upper =
+          index < sweepLineStatus.length - 1 ? sweepLineStatus[index + 1] : null
+        sweepLineStatus = sweepLineStatus.filter((s) => s !== segment)
+        if (lower && upper)
+          handlePotentialIntersection(lower, upper, events, intersections)
+      })
+    }
+
+    if (event.types.includes('intersection')) {
+      // Re-check all pairs of segments at this x-coordinate for intersections
```

Comment:
This also suffers from poor time complexity: First we do a linear filtering, then we do a quadratic brute force finding of intersections.